JuliaSyntax parser-based REPL completions overhaul

xal-0 opened the pull request:
# Overview

As we add REPL features, bugs related to the ad-hoc parsing done by
`REPLCompletions.completions` have crept in.  This pull request replaces most of
the manual parsing (regex, `find_start_brace`) with a new approach that parses
the entire input buffer once, before and after the cursor, using JuliaSyntax.
We then query the parsed syntax tree to determine the kind of completion
to be done.

# Changes
- New, JuliaSyntax-based completions mechanism.

- The `complete_line` interface now has the option of replacing arbitrary
  regions of text in the input buffer by returning a `Region` (`Pair{Int, Int}`
  for consistency with the convention in LineEdit, and `pos` being a 0-based
  byte offset).  ~~This is used to improve the handling of auto-inserted closing
  delimiters.~~. Leaving this unchanged for now.

- Fixes parsing-related bugs:
  - fix #55420
  - fix #55429
  - fix #55518
  - fix #55520
  - fix #55842
  - fix #56389
  - fix #57307
  - fix #57611
  - fix #57624
  - fix #58099

- Fixes some bugs that exist on 28d3bd56d36 that were found by fuzzing:
  - `x \"` + `TAB` throws a `FieldError` exception
  - String completion would sometimes delete the entire input buffer.
  - Completions should not happen inside comments.

- The duplicate code for path completion in strings, `Cmd`-strings, and the
  shell has been removed, causing paths to complete the same way for all three.
  Now, `~` is expanded in two situations:
  - If `foo` exists, or if `foo` does not exist but there are no
    possible completions:
    ```
    "~/foo/b|"     =TAB=>   "~/foo/bar|"
    "~/foo/bar|"   =TAB=>   "/home/user/foo/bar|"
       OR
    "~/foo/bar"|   =TAB=>   "/home/user/foo/bar"|
    ```

  - If the current path ends with a `/` and you hit TAB again:
    ```
    "~/foo/|"      =TAB=>   "/home/user/foo/|"
       OR
    "~/foo/"|      =TAB=>   "/home/user/foo/"|
    ```

# Future work
- Method completions could be changed to look for methods with exactly the given
  number of arguments if the closing `)` is present, and search for signatures
  with the right prefix otherwise.

    - It would be nice to be able to search by type as well as value (perhaps
      by putting `::T` in place of arguments).

- Other REPL features could benefit from JuliaSyntax, so it might be worth
  sharing the parse tree between completions and other features:

    - Emacs-style sexpr navigation: `C-M-f`/`C-M-b`/`C-M-u`, etc.
    - Improved auto-indent.

- It would be nice if hints worked even when the cursor is between text.

- `CursorNode` is a slightly tweaked copy of `SyntaxNode` from JuliaSyntax that
  tracks the parent node but includes all trivia.  It is used with `seek_pos`,
  which navigates to the innermost node at a given position so we can examine
  nearby nodes and the parent.  This could probably duplicate less code from
  JuliaSyntax.

xal-0 committed:
REPL: Allow completions to replace arbitrary regions of text

Adds another permitted return type for complete_line, where the second element
of the tuple is a Region (a Pair{Int, Int}) describing the region of text to be
replaced.  This is useful for making completions work consistently when the
closing delimiter may or may not be present: the cursor can be made to "jump"
out of the delimiters regardless of whether it is there already.

  "exam|    =TAB=>   "example.jl"|
  "exam|"   =TAB=>   "example.jl"|

xal-0 committed:
REPL: JuliaSyntax completions overhaul

This commit replaces the heuristic parsing done by REPLCompletions.completions
with a new approach that parses the entire input buffer once with JuliaSyntax.
In addition to fixing bugs, the more precise parsing should allow for new
features in the future.

Some features now work in more situations "for free", like dictionary key
completion (the expression evaluated to find the keys is now more precise) and
method suggestions (arguments beyond the cursor can be used to narrow the list).

The tests have been updated to reflect slightly differing behaviour for string
and Cmd-string completion: the new code returns a character range encompassing
the entire string when completing paths (not observable by the user), and the
behaviour of '~'-expansion has be tweaked to be consistent across all places
where paths can be completed.  Some escaping issues have also been fixed.

Fixes: #55420, #55518, #55520, #55842, #56389, #57611

IanButterworth commented:
This sounds great! I think we should backport it to at least 1.12 given it fixes so many issues.

giordano commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -2484,3 +2494,43 @@ let (c, r, res) = test_complete_context("global xxx::Number = Base.", Main)
+    @test s[r] == "type_"
+end
+
+# #55518
+let s = "CompletionFoo.@barfoo kwtest"
+    c, r = test_complete(s)
+    @test isempty(c)
+end
```
I'm not sure this quite cover all the broken cases reported in #55518 (there are a few scattered in the discussion)

xal-0 replied:
I've added a few more from the thread

giordano reviewed (commented):
@xal-0 first of all, speaking as the author of most of the issues linked to this PR, thank you for tackling this!

Also, can you please add a test for #57772 as well?  I verified this works already (which is amazing!)

StefanKarpinski commented:
Bravo. Truly excellent quality-of-life PR 👏🏻

giordano commented:
> Bravo. Truly excellent quality-of-life PR 👏🏻

Best part is that this PR has a negative diff, depsite the fact it added many tests.

giordano commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -158,6 +158,10 @@ let ex =
             exported_symbol(::WeirdNames) = nothing
 
+            macro ignoremacro(e...)
+                :nothing
```
Should this be
```suggestion
                nothing
```
?
```julia-repl
julia> macro ignoremacro(e...)
           :nothing
       end
@ignoremacro (macro with 1 method)

julia> @macroexpand @ignoremacro
:(Main.nothing)

julia> macro ignoremacro(e...)
           nothing
       end
@ignoremacro (macro with 1 method)

julia> @macroexpand @ignoremacro

julia> 
```

xal-0 replied:
I was actually unaware that so many things were self-quoting!

xal-0 committed:
REPL: Add tests for #55420, #55429, #55842, #57307, #57624

Also fix test failing for silly reason

xal-0 committed:
Add test for #57772

xal-0 committed:
REPL: Add additional tests from #55518 discussion

xal-0 committed:
REPL: Avoid triggering method completion when cursor on function name

Since we parse the entire input now, we need to avoid triggering method
completion when the cursor is on the function name of a valid call.

IanButterworth commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -2484,3 +2502,109 @@ let (c, r, res) = test_complete_context("global xxx::Number = Base.", Main)
+end
+
+# #56389
+let s = "begin\n  using Linear"
+    c, r = test_complete(s)
+    @test "LinearAlgebra" in c
+    @test r == 15:20
+    @test s[r] == "Linear"
+end
```
This doesn't pass when run via `Pkg.test("REPL")` because `Base.runtests` puts all stdlibs on the load path, whereas `Pkg.test` only does REPL deps and test deps.
```suggestion
let s = "begin\n  using Ran"
    c, r = test_complete(s)
    @test "Random" in c
    @test r == 15:17
    @test s[r] == "Ran"
end
```

IanButterworth commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -1550,44 +1326,52 @@ function shell_completions(string, pos, hint::Bool=false)
+
+    # Expand '~' if the user hits TAB after exhausting completions (either
+    # because we have found an existing file, or there is no such file).
+    full_path = ispath(path) || isempty(paths)
```
I found two cases that can cause `ispath` to error, so I think we should make a `safe_ispath`

Something like this? Or just never throw and just return false?
```
function safe_ispath(path::AbstractString)
    try
        return ispath(path)
    catch err
        if err isa Base.IOError # i.e. name too long (ENAMETOOLONG)
            return false
        elseif err isa Base.ArgumentError && occursin("embedded NULs", err.msg)
            return false
        else
            rethrow()
        end
    end
end
```

xal-0 replied:
I'll avoid modifying Base for the time being because we already convert some `access(2)` return values to `false`.  It feels a little hacky to keep adding exceptions here. https://github.com/JuliaLang/julia/blob/bc98abcfbe392d2a0a09ed631b85f4675a34b380/base/stat.jl#L376-L379

Of the errors that could be returned for `F_OK` that we might want to handle, it looks like we're missing `EACCES`, `EIO`, `ELOOP`, and `ENAMETOOLONG`.

xal-0 committed:
Update stdlib/REPL/test/replcompletions.jl

Co-authored-by: Ian Butterworth <i.r.butterworth@gmail.com>

xal-0 committed:
REPL: Catch IOError, ArgumentError when calling ispath

Co-authored-by: Ian Butterworth <i.r.butterworth@gmail.com>

xal-0 committed:
Update stdlib/REPL/test/replcompletions.jl

Co-authored-by: Mosè Giordano <765740+giordano@users.noreply.github.com>

xal-0 committed:
Merge branch 'master' into juliasyntax-repl

xal-0 committed:
Merge branch 'master' into juliasyntax-repl

xal-0 commented:
`isdir` suffers from similar problems as `ispath`. On Windows, `isdir("D:\\")` can fail if no disc is inserted, because libuv doesn't handle `ERROR_NOT_READY`:

```
julia> isdir("D:\\")
ERROR: IOError: stat("D:\\"): unknown error (UNKNOWN)
Stacktrace:
 [1] uv_error
   @ .\libuv.jl:106 [inlined]
 [2] checkstat
   @ .\stat.jl:186 [inlined]
 [3] stat(path::String)
   @ Base.Filesystem .\stat.jl:192
 [4] isdir(path::String)
   @ Base.Filesystem .\stat.jl:586
 [5] top-level scope
   @ REPL[1]:1
```

xal-0 commented:
I am going to disable the shell completion tests until the shell mode can parse Windows paths...
```
shell> cd C:\Users
ERROR: IOError: cd("C:Users"): no such file or directory (ENOENT)
Stacktrace:
 [1] uv_error
   @ .\libuv.jl:106 [inlined]
 [2] cd(dir::String)
   @ Base.Filesystem .\file.jl:91
 [3] repl_cmd(cmd::Cmd, out::Base.TTY)
   @ Base .\client.jl:64
 [4] top-level scope
   @ none:1
```

xal-0 committed:
REPL: Escape path completions after joining dir and path

Also cleans up do_cmd_escape, so that it can use different escaping syntax from
the shell mode (which we may want to make similar to cmd.exe on Windows).

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into juliasyntax-repl

xal-0 committed:
REPL: new backslash escape hack for Windows Pkg completions

vtjnash commented:
> until the shell mode can parse Windows paths

The shell mode has no problem with Windows paths which is why the REPL has tests for it. It just sounds like you lost a call to `Base.shell_escape` on that code path.

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -302,16 +304,23 @@ const sorted_keyvals = ["false", "true"]
-    # escape_raw_string with delim='`' and ignoring the rule for the ending \
-    return replace(s, r"(\\+)`" => s"\1\\`")
+function do_cmd_escape(s)
+    return Base.shell_escape_posixly(Base.escape_raw_string(s, '`'))
```
This seems like nonsense: I can't think of any case where we'd expect it'd be useful for the the REPL to pass the string unmodified to a posix shell for use as an argument in a julia script, which is what this transform order implements.

The previous transform took a string `text` that was already in posix-shell form (e.g. for raw input to shell_parse for `shell> text` mode) and corrects it for the julia parser (e.g. for ``:(`text`)``)
```
julia> a = `\\ b\\\\`
`'\' 'b\'`
```

xal-0 replied:
I intended `cmd_escape` to make it possible to complete shell commands inside Cmd-strings, but mixed up the order (``Base.escape_raw_string(Base.shell_escape_posixly(s), '`')`` does what I want I think):
```
julia> println(R.do_cmd_escape("file ` 1"))
'file \` 1'

julia> @macroexpand `'file \` 1'`
:(Base.cmd_gen((("file ` 1",),)))
```

xal-0 replied:
Completion with `shell_escape_posixly` is still unwieldy because you get partial completions that insert only the opening quote, and can't complete any further.

xal-0 committed:
Update stdlib/REPL/src/REPLCompletions.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

xal-0 commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -1136,399 +957,352 @@ function get_loading_candidates(pkgstarts::String, project_file::String)
-    for name in get_loading_candidates(pkgstarts, project_file)
-        push!(suggestions, PackageCompletion(name))
+function complete_loading_candidates!(suggestions::Vector{Completion}, s::String)
+    for name in ("Core", "Base")
```
Do you have an example where it misbehaves? This should only fire when we're completing the first component of an import path with no leading dots.

vtjnash replied:
Oh, right, I switched the argument order of startswith in my head and was just confused

xal-0 committed:
Use '/' as directory separator for shell completions

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into juliasyntax-repl

xal-0 committed:
Restore previous Pkg.jl path completion

xal-0 committed:
Merge remote-tracking branch 'upstream/master' into juliasyntax-repl

xal-0 committed:
Allow REPL completions inside string interpolation

IanButterworth commented:
What's the status of this? I'm just wondering whether it's worth getting it in and continuing work in PRs?

xal-0 commented:
With the REPL completions working well enough on Windows again, this is ready for review again. 👍

KristofferC commented:
🎤 ⬇️

IanButterworth commented:
As someone who tried and failed to fix the issues this closes by patching the old approach, thank you for fixing this properly!
